### ISSUE TITLE
"reverse" for wifi condition not working

### ISSUE THREAD

**Issue by renyuneyun** [renyuneyun/Easer#377]
@renyuneyun thank you for adding this feature.

i installed the beta release to test this out. in order to toggle a VPN on when wireguard is in the background, you must set wireguard to "not optimize" for battery usage. with this set, i can trigger wireguard enable when eg. "WiFi enabled" is false.

one problem i ran into: i'd like to enable VPN only when **not** on home WiFi. however, the "reverse scenario" for WiFi connectivity appears broken. the condition never triggers, though the positive condition does.

i see you tried to address this in a previous PR https://github.com/renyuneyun/Easer/commit/3331d91331efe32cabdaa73c8d51b6e0e2e2508e

_Originally posted by @khimaros in https://github.com/renyuneyun/Easer/issues/361#issuecomment-838582344_

**Comment by renyuneyun** [comment]
@khimaros just to check: are you using Condition or Event for the wifi connection?

**Comment by khimaros** [comment]
@renyuneyun hello! i am using condition for this script.

**Comment by renyuneyun** [comment]
That's strange... My test works as expected...
Would you mind sharing your relevant settings (scripts, conditions, etc)? Maybe something can be found from there.

May 15, 2021 00:10:12 khimaros ***@***.***>:

**Comment by khimaros** [comment]
@renyuneyun -- i believe i see what's happening.

the way i was testing this was by disabling WiFi entirely (via quick toggle); in that case, the "Home WiFi" reverse scenario condition does not trigger. however, "Disconnect" in the WiFi settings or connecting to a non-matching network does trigger it.

the workaround is to add another condition of "Connectivity Type: WiFi" with another script using that as the reverse scenario to also enable the VPN. so there will be three scripts total.

i am using Easer `0.8.2-beta` from F-Droid on an AOSP build (Pixel 3).

for reference, i've included the original configuration and the workaround below.

### original configuration

#### condition

```
{"name":"Home WiFi","version":16,"condition":{"spec":"wifi connection","data":"{\"essid\":[\"<ELIDED>\"]}"}}
```

#### profile

```
{"name":"Disable VPN","version":16,"operation":[{"spec":"wireguard","data":"{\"tunnel_name\":\"<ELIDED>\",\"tunnel_state\":false}"}]}
{"name":"Enable VPN","version":16,"operation":[{"spec":"wireguard","data":"{\"tunnel_name\":\"<ELIDED>\",\"tunnel_state\":true}"}]}
```

#### script

```
{"name":"Disable VPN home","version":16,"active":true,"profile":"Disable VPN","trigger":{"type":"condition","condition":"Home WiFi"},"reverse":false}
{"name":"Enable VPN away","version":16,"active":true,"profile":"Enable VPN","trigger":{"type":"condition","condition":"Home WiFi"},"reverse":true}
```

### workaround configuration

#### condition

```
{"name":"Home WiFi","version":16,"condition":{"spec":"wifi connection","data":"{\"essid\":[\"<ELIDED>\"]}"}}
{"name":"WiFi Connected","version":16,"condition":{"spec":"connectivity","data":"[10]"}}
```

#### profile

```
{"name":"Disable VPN","version":16,"operation":[{"spec":"wireguard","data":"{\"tunnel_name\":\"<ELIDED>\",\"tunnel_state\":false}"}]}
{"name":"Enable VPN","version":16,"operation":[{"spec":"wireguard","data":"{\"tunnel_name\":\"<ELIDED>\",\"tunnel_state\":true}"}]}
```

#### script

```
{"name":"Disable VPN home","version":16,"active":true,"profile":"Disable VPN","trigger":{"type":"condition","condition":"Home WiFi"},"reverse":false}
{"name":"Enable VPN away","version":16,"active":true,"profile":"Enable VPN","trigger":{"type":"condition","condition":"Home WiFi"},"reverse":true}
{"name":"Enable VPN mobile","version":16,"active":true,"profile":"Enable VPN","trigger":{"type":"condition","condition":"WiFi Connected"},"reverse":true}
```

**Comment by huuhaa** [comment]
Just douplr checking... Since at begin this is marked as bug and is still open.. Is this known bug or should following work:

-Script VPN on demand > reverse for condition home wifi SSID > profile: turn on vpn
-Child script  of "VPN on demand" > Wifi disabled > event (not reversed) "wifi is enabled > no" > profile: turn vpn on

At profile:
"turn vpn on" > wireguard ticked > Tunnel Name given > Tunnel State: Up

Does that seem ok?

Also, is location really needed for this app, would think wifi status etc doesn't require location. But on the other hand at android also BT is under location permission.. But while location enabled, it actually seems to get GPS location...

**Comment by Xerono** [comment]
I have a problem that may be related - Easer doesn't notice when I'm on wifi. It's the same case as in the opening post here, but without the "reverse". "Reverse" works fine for me (Wireguard gets enabled), but I can't get it to disable once I'm back in my home wifi. This happens everytime.

In addition, I have the same problem as teemue, as in I sometimes have to toggle Wireguard manually before I get a connection. I can't reproduce this behaviour, it's random. I assumed the problem came from being not connected to a mobile network before setting up the tunnel, but even if I only activate wireguard once a connection exists, the problem can occur.

**Comment by madpsy** [comment]
@Xerono Does the VPN disconnect if you hit the refresh icon on the Outline pane? I had the same issue and ended up duplicating the script but using a Condition instead of Event and have to have both versions active.

**Comment by Xerono** [comment]
@madpsy Yes, refresh seems to change to the correct script. I tried using events and conditions, can you post your setup?

**Comment by prontog** [comment]
Hello and thanks for this very useful app. I just wanted to thank @khimaros for his comment that helped me troubleshoot the "Wireguard Tunnel State: Up" profile. I had to disable battery optimization for this to work. The weird part is that "Wireguard Tunnel State: Down" profile always worked. Thanks again.

**Comment by renyuneyun** [comment]
Hi. Sorry for not being able to find the cause, as I couldn't reproduce this on my phone and was busy during the period... (Happy to accept PRs if this could be automatically tested.)

There is a recent report #430, which seems to be related to this issue. Could I have a quick check from people: does the issue still exist on v0.8.2.3-beta2?
